include/chainbase/chainrocks.hpp

```diff
@@ -0,0 +1,818 @@
+      /// Note that the session be explicitly told not to do so, by
+      /// calling the method `push`.
+      class session {
+      public:
+         session(const session&) = delete;
+         session& operator= (const session&) = delete;
+
+         session(session&& s)
+            : _index{s._index}
+            , _apply{s._apply}
+         {
+            s._apply = false;
+         }
+
+         session& operator= (session&& s) {
+            if (this == &s) {
+               return *this;
+            }
+      
+            if (_apply) {
+               _index.undo();
+            }
+         
+            _apply = s._apply;
+            s._apply = false;
+            return *this;
+         }
+
+         /// RAII functionality; upon the destruction of `session` it
+         /// shall be determined whether or not the `_index` is acted
+         /// upon by `undo`.
+         ~session() {
+            if(_apply) {
+               _index.undo();
+            }
+         }
+
+         void push() {
+            _apply = false;
+         }
+
+         void undo() {
+            if (_apply) {
+               _index.undo();
+               _apply = false;
+            }
+         }
+
+         void squash() {
+            if (_apply) {
+               _index.squash();
+               _apply = false;
+            }
+         }
+
+         int64_t revision() const {
+            return _revision;
+         }
+
+      private:
+         friend class index;
+
+         /// Note that it is not possible to directly construct a
+         /// session object because this responsibility shall be left
+         /// to the method `start_undo_session`.
+         session(index& idx, int64_t revision)
+            : _index{idx}
+            , _revision{revision}
+         {
+            if(_revision == -1) {
+               _apply = false;
+            }
+         }
+
+         /// The given `index` for this session to hold.
+         index& _index;
+
+         /// The predicate determining whether or not this `session`
+         /// is elligle to be acted upon by either `undo` or `squash`.
+         bool _apply{true};
+
+         /// The unique revision number held by each `session` object.
+         /// Note that this `_revision` number and the `_revision`
+         /// number of the held `index` will be exactly identical to
+         /// eachother.
+         int64_t _revision{};
+      };
+
+      /// Probably the most interested method in the chainbase
+      /// repository. It can be essentially thought of as one of two
+      /// ideas: the start of a block (containing transactions) in a
+      /// blockchain, or the start of a transaction (containing
+      /// actions) within a block. Each of which have the ability to
+      /// `commit` if valid or to `undo` if invalid due to the nature
+      /// of these data structures.
+      session start_undo_session(bool enabled) {
+         if (enabled) {
+            _stack.emplace_back(undo_state{});
```

Comment:
Sufficient to just write: `_stack.emplace_back();`.

Reply:
Resolved in [47a9f46f3d8cc8e8948faddeef56e2e06aba5a7f](https://github.com/EOSIO/chainbase/blob/47a9f46f3d8cc8e8948faddeef56e2e06aba5a7f/include/chainbase/chainrocks.hpp#L445-L460).